### ISSUE TITLE
Roadmap to 1.0?

### ISSUE THREAD

**Issue by AnthonyMastreanCcsi** [dotnet/BenchmarkDotNet#2026]
We use BenchmarkDotNet in an FDA-regulated environment to design and develop a medical device. We have to account for all of the off-the-shelf (OTS) components. We are responsible for understanding the development process, maintenance and support lifecycle, design limitations, and other quality properties of every OTS components we use.

The lack of a major version and a roadmap does not prevent use of a component, but it introduces a documentation burden to explain or justify the use of a prerelease component.

It would be awesome to see GitHub [releases](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases) and [milestones](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones), at least, and maybe a ["roadmap" project](https://docs.github.com/en/issues/trying-out-the-new-projects-experience/about-projects).

**Comment by AnthonyMastrean** [comment]
(This was me, from another account that I meant to delete... I have sync'ed my email addresses 😳😅)

**Comment by AndreyAkinshin** [comment]
> I see that every version has a changelog entry on the [website](https://benchmarkdotnet.org/changelog/index.html), but not a GitHub "release" which provides a useful, albeit hidden, Atom feed link (this is useful for our configuration management practice).

It's a valid point, I have created a separate issue for that: #2036. I will think about ways to automate the pushing of changelogs to the GitHub releases (I don't want to do it manually each time, we are trying to completely automate the release process which is half-manual right now).

> Roadmap to 1.0

I'm not completely satisfied with our statistical engine. While it's quite mature and stable, there are some corner cases that may cause problems in measurements and result presentations. I think that 1.0 should be released once the new statistical engine is ready. Unfortunately, it's a huge research project without ETA (you can find research notes in my [blog](https://aakinshin.net/posts/)).

Regardless of the fact that 1.0 is not published yet, the current BenchmarkDotNet version is stable and [widely-adopted](https://github.com/dotnet/BenchmarkDotNet#who-uses-benchmarkdotnet). I believe that the lack of the official 1.0 version shouldn't a blocker to adopt BenchmarkDotNet in your project.

**Comment by AnthonyMastrean** [comment]
It's all good! I just need to provide more thorough documentation. The guides on the website are actually very useful in providing evidence of quality (good practices, how it works, etc.).

https://benchmarkdotnet.org/articles/guides/how-it-works.html

**Comment by timcassell** [comment]
@AndreyAkinshin Before 1.0, I was thinking some of the public types that are only public for the generated projects to consume could be made internal, and add `[InternalsVisibleTo]` so the generated projects can still use them. What do you think?

[Edit] I'm not sure if that would work for the InProcessEmitToolchain, but I think so.

**Comment by YegorStepanov** [comment]
Please, do not silently release 1.0 without an announcement.

Topics that I want to cover during the year:
- BDN should not throw unhandled exceptions.
- Unify toolchains. (for example, `InProcessEmitToolchain` supports `[Params("111")] public int Field;`)
- Move different checks to validators (to simplify code duplication with analyzers).
- Add analyzers.
- Try to replace `RPlots` to the C# plots ([ScottPlot 5](https://github.com/ScottPlot/ScottPlot) is the only C# library that fits BDN, I haven't experimented with it yet. I'll write a prototype).
- Rewrite the documentation: from the sample-based docs to the normal :)

**Comment by filzrev** [comment]
Redesigning the command-line arguments is also a candidate before releasing version 1.0.
See: https://github.com/dotnet/BenchmarkDotNet/pull/3162#issuecomment-4592633337
